pkg/operator/sapinstancestart_v1_test.go

```diff
@@ -0,0 +1,417 @@
+	ctx := context.Background()
+
+	gray := sapcontrol.STATECOLORSAPControlGRAY
+
+	planGetProcesses := suite.mockSapcontrol.
+		On("GetProcessListContext", ctx, mock.Anything).
+		Return(
+			&sapcontrol.GetProcessListResponse{
+				Processes: []*sapcontrol.OSProcess{
+					{
+						Dispstatus: &gray,
+					},
+				},
+			}, nil,
+		).
+		Once()
+
+	verifyGetProcesses := suite.mockSapcontrol.
+		On("GetProcessListContext", mock.Anything, mock.Anything).
+		Return(nil, errors.New("error getting processes in verify")).
+		Once().
+		NotBefore(planGetProcesses)
+
+	suite.mockSapcontrol.
+		On("StartContext", ctx, mock.Anything).
+		Return(nil, nil).
+		NotBefore(planGetProcesses).
+		On("StopContext", ctx, mock.Anything).
+		Return(nil, nil).
+		NotBefore(verifyGetProcesses).
+		On("GetProcessListContext", mock.Anything, mock.Anything).
+		Return(
+			&sapcontrol.GetProcessListResponse{
+				Processes: []*sapcontrol.OSProcess{
+					{
+						Dispstatus: &gray,
+					},
+				},
+			}, nil,
+		).
+		NotBefore(verifyGetProcesses)
+
+	sapInstanceStartOperator := operator.NewSAPInstanceStart(
+		operator.OperatorArguments{
+			"instance_number": "00",
+		},
+		"test-op",
+		operator.OperatorOptions[operator.SAPInstanceStart]{
+			OperatorOptions: []operator.Option[operator.SAPInstanceStart]{
+				operator.Option[operator.SAPInstanceStart](operator.WithCustomStartSapcontrol(suite.mockSapcontrol)),
+				operator.Option[operator.SAPInstanceStart](operator.WithCustomStartInitialDelay(0 * time.Second)),
+			},
+		},
+	)
+
+	report := sapInstanceStartOperator.Run(ctx)
+
+	suite.Nil(report.Success)
+	suite.Equal(operator.VERIFY, report.Error.ErrorPhase)
+	suite.EqualValues("verify instance started failed: error getting instance process list: error getting processes in verify", report.Error.Message)
+}
+
+func (suite *SAPInstanceStartOperatorTestSuite) TestSAPInstanceStartVerifyTimeout() {
+	ctx := context.Background()
+
+	gray := sapcontrol.STATECOLORSAPControlGRAY
+
+	suite.mockSapcontrol.
+		On("GetProcessListContext", mock.Anything, mock.Anything).
+		Return(
+			&sapcontrol.GetProcessListResponse{
+				Processes: []*sapcontrol.OSProcess{
+					{
+						Dispstatus: &gray,
+					},
+				},
+			}, nil,
+		).
+		Times(3).
+		On("StartContext", ctx, mock.Anything).
+		Return(nil, nil).
+		On("StopContext", ctx, mock.Anything).
+		Return(nil, nil)
+
+	sapInstanceStartOperator := operator.NewSAPInstanceStart(
+		operator.OperatorArguments{
+			"instance_number": "00",
+			"timeout":         0.0,
+		},
+		"test-op",
+		operator.OperatorOptions[operator.SAPInstanceStart]{
+			OperatorOptions: []operator.Option[operator.SAPInstanceStart]{
+				operator.Option[operator.SAPInstanceStart](operator.WithCustomStartSapcontrol(suite.mockSapcontrol)),
+				operator.Option[operator.SAPInstanceStart](operator.WithCustomStartInitialDelay(0 * time.Second)),
+				operator.Option[operator.SAPInstanceStart](operator.WithCustomStartInterval(0 * time.Second)),
```

Comment:
I have doubts about 0 seconds timeouts/intervals, wondering whether non-zero values, very low ones tho, might help making the test more realistic.

However, I understand the challenges of these kind of test.

Reply:
Yeah...
What a pain hehe
It is never easy with timeout and intervals.
I didn't want to test that `context.WithTimeout` works, because I'm sure it does, and adding real times doesn't change much (you run the `GetProcessList` more times).